### ISSUE TITLE
Cannot get weights from get_embedding_layer 

### ISSUE THREAD

**Issue by kaushalshetty** [RaRe-Technologies/gensim#1564]
So basically i am training a doc2vec model. I came across something called get_embedding_layer in KeyedVectors class which returns a embedding layer. But I am unable to retrieve the weights for this layer. 
The following is my code:
```
text_corpora = [["hi how are you"],["I am all good"],["We are all good"]]
sentences = [TaggedDocument(words= text_corpora[0][0].split(),tags = ['1']),TaggedDocument(words= text_corpora[1][0].split(),tags = ['2']),TaggedDocument(words= text_corpora[2][0].split(),tags = ['3'])]
model = Doc2Vec(size = 200, window = 300, min_count = 1, workers = 4)
model.build_vocab(sentences)

model.train(sentences,total_examples=model.corpus_count,epochs=model.iter)
```
`model.wv.get_embedding_layer().get_weights()`  should give me the embeddings but its returning an empty array.
` Is model.wv.syn0` a way to retrieve the weights and initialize in the embedding layer in keras ?

Also please say whether this is the correct way to train the model.

**Comment by kaushalshetty** [comment]
Also please say whether this is the correct way to train a word2vec model.

**Comment by chinmayapancholi13** [comment]
@kaushalshetty Yes you are training the model correctly here. `get_embedding_layer` returns a Keras `Embedding` object which you can use in your Keras model. To get the appropriate value from `get_weights()`, you would have to train your Keras model first.

The following code works for me: 

``` python
from gensim.models import Doc2Vec
from gensim.models.doc2vec import TaggedDocument

from keras.models import Sequential

text_corpora = [["hi how are you"],["I am all good"],["We are all good"]]
sentences = [TaggedDocument(words= text_corpora[0][0].split(),tags = ['1']),TaggedDocument(words= text_corpora[1][0].split(),tags = ['2']),TaggedDocument(words= text_corpora[2][0].split(),tags = ['3'])]

model = Doc2Vec(size = 200, window = 300, min_count = 1, workers = 4)
model.build_vocab(sentences)
model.train(sentences,total_examples=model.corpus_count,epochs=model.iter)

emb1 = model.wv.get_embedding_layer()
kmodel = Sequential()
kmodel.add(emb1)
kmodel.compile('rmsprop', 'mse')
print(emb1.get_weights())

```

My output: 
```
[array([[  5.70312084e-04,   1.37853285e-03,   1.10545720e-03, ...,
         -2.09680595e-03,   2.03875126e-03,   2.18590396e-03],
       [ -5.36838721e-04,   1.75698334e-03,   4.77984751e-04, ...,
          5.00339782e-04,   2.20681145e-03,  -1.06832071e-03],
       [ -3.36936209e-04,   2.00897851e-03,  -1.53642369e-03, ...,
          2.05683638e-03,   1.46236166e-03,   2.35496485e-03],
       ..., 
       [  1.77507254e-03,  -9.03652923e-04,   2.23973207e-03, ...,
          1.04890577e-03,   1.18979590e-03,   9.87872991e-05],
       [ -1.52916857e-03,  -6.19521074e-04,   1.72100205e-04, ...,
         -2.42557464e-04,   2.24056025e-03,  -3.86744039e-04],
       [  2.34267348e-03,  -1.35466413e-04,  -1.98745239e-03, ...,
          9.76303883e-04,  -1.53724244e-03,  -2.97894323e-04]], dtype=float32)]
```
Let me know if this doesn't sort out your problem. :)

**Comment by kaushalshetty** [comment]
Got it. Thank you

**Comment by kaushalshetty** [comment]
@chinmayapancholi13 I have one more doubt. I have trained my model on a list of taggeddocuments.
Initially my window_size was 8 and i got some document vector say X. After that I changed my window size to 20 but my document vector is still X for a particular document. I suppose the document vectors must change w.r.t the window size right? 

Also I wanted to find the most_similar documents among the set of documents. So i initialized a sting with one of the document in my dataset itself and tokenized it. After which i did a infer_vector  followed by most_similar. But most_similar should actually have returned me with the same document (since i am inferring the same vector) at least with a good similarity if not 1. Its not even coming in the top20. I have trained the model for 100 epochs but no hopes.

**Comment by kaushalshetty** [comment]
@chinmayapancholi13 

Also I wanted to find the most_similar documents among the set of documents. So i initialized a sting with one of the document in my dataset itself and tokenized it. After which i did a infer_vector followed by most_similar. But most_similar should actually have returned me with the same document (since i am inferring the same vector) at least with a good similarity if not 1. Its not even coming in the top20. I have trained the model for 100 epochs but no hopes.

**Comment by kaushalshetty** [comment]
On top of that the most_similar property is not making any sense.

**Comment by chinmayapancholi13** [comment]
@kaushalshetty Results obtained for `most_similar` would depend on the training params and corpus size. You can check out issue https://github.com/RaRe-Technologies/gensim/issues/1560 and see if the discussion there helps.

**Comment by piskvorky** [comment]
@kaushalshetty `On top of that the most_similar property is not making any sense.` -- most_similar should be a method, not a property. Or what do you mean by this?

The similarity between infer_vector vs trained vectors should indeed be high, your expectations is correct. Try increasing the convergence parameter `step` -- does that help?

**Comment by kaushalshetty** [comment]
@piskvorky Sorry for the misnomer. Yes i meant the method itself. 
I am not sure where I can set the `step` parameter.

**Comment by piskvorky** [comment]
I meant the `steps` parameter to `infer_vector(steps=N)` here: https://radimrehurek.com/gensim/models/doc2vec.html#gensim.models.doc2vec.Doc2Vec.infer_vector

Tuning: the main params that affect quality include the preprocessing (=how you get nice, normalized tokens even before doc2vec starts) and the amount and quality of the training data (=number of documents, number of training epochs).

As for the other tweak-able params (training modes, learning rates, dimensionality, window size) that's usually domain and goal specific. See @chinmayapancholi13 's link above. But having clean data and a good evaluation metric typically trumps micro-tweaks.

**Comment by kaushalshetty** [comment]
Cool. I will tweak the above parameters and see. Thanks a lot . @chinmayapancholi13  @piskvorky
